adaptdl/adaptdl/torch/data.py

```diff
@@ -405,11 +475,13 @@ def __iter__(self):
             while not done:
                 self.sampler.set_epoch(epoch, index=self._elastic.current_index)  # noqa: E501
                 self.batch_sampler.batch_size = self._elastic._sync_local_bsz()
+                iterations = self._elastic.get_iterations(len(self.sampler))
                 for idx, batch in enumerate(super().__iter__()):
+                    if idx > iterations:
+                        break
                     with self._elastic.profile():
                         yield batch
-                        # Increment by the number of data samples processed
-                        self._elastic.current_index += self.current_batch_size  # noqa: E501
+                        self._elastic.increment_index()
```

Comment:
Is it possible to remove all changes to this class? Otherwise, you'll need to also make sure the BPTTIterator works and test the transformer. And this complexity will extend to all future data loading classes

Reply:
I don't think that is possible, at least not without creating a layer on top of elastic sampler. The primary issue is that the length of the dataset really needs to be zero mod accumulation steps. I think BPTTIterator will need to be changed for it to work with gradient accumulation, but it should be a one line change.